pkgs/build-support/trivial-builders/default.nix

```diff
@@ -625,18 +625,21 @@ rec {
 
   # Docs in doc/build-helpers/trivial-build-helpers.chapter.md
   # See https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-writeReferencesToFile
-  writeReferencesToFile = path: runCommand "runtime-deps"
+  # TODO: Convert to throw after Nixpkgs 24.05 branch-off.
+  writeReferencesToFile = (if config.allowAliases then lib.warn else throw)
+    "writeReferencesToFile is deprecated in favour of writeReferenceClosureToFile"
+    (path: writeReferenceClosureToFile [ path ]);
+
+  # Docs in doc/build-helpers/trivial-build-helpers.chapter.md
+  # See https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-writeReferenceClosureToFile
+  writeReferenceClosureToFile = paths: runCommand "runtime-deps"
     {
-      exportReferencesGraph = [ "graph" path ];
+      __structuredAttrs = true; # Get the cleaner exportReferencesGraph interface
+      exportReferencesGraph.graph = paths;
+      nativeBuildInputs = [ jq ];
     }
     ''
-      touch $out
-      while read path; do
-        echo $path >> $out
-        read dummy
-        read nrRefs
-        for ((i = 0; i < nrRefs; i++)); do read ref; done
-      done < graph
+      jq -r ".graph | sort_by(.path) | .[].path" ./.attrs.json > "$out"
```

Comment:
Yaaay this looks so simple now!

Nit: looking at https://nixos.org/manual/nix/unstable/language/advanced-attributes.html?highlight=__structured#adv-attr-structuredAttrs, I guess the "public" door is `$NIX_ATTRS_JSON_FILE` rather than the path itself.